src/daft-scan/src/scan_task_iters/retrieve_parquet_metadata.rs

```diff
@@ -0,0 +1,274 @@
+use std::{
+    collections::{HashMap, VecDeque},
+    sync::Arc,
+};
+
+use common_error::DaftResult;
+use common_file_formats::ParquetSourceConfig;
+use common_runtime::get_io_runtime;
+use daft_io::{get_io_client, IOConfig, IOStatsContext};
+use daft_parquet::read::read_parquet_metadata;
+use futures::future::try_join_all;
+use itertools::Itertools;
+use parquet2::metadata::FileMetaData;
+
+/// Performs Parquet metadata retrieval for the provided iterator of [`NeedsParquetMetadata`]s
+///
+/// This will transform each [`NeedsParquetMetadata`] into its specified [`NeedsParquetMetadata::WithMetadata`] by
+/// performing batched Parquet metadata retrieval whenever [`NeedsParquetMetadata::should_fetch`] is ``true``.
+pub fn batched_parquet_metadata_retrieval<T>(
```

Comment:
I would favor builders on the struct rather than these global functions.

```
FetchParquetMetadataByWindows::new(inputs, max_retrieval_batch_size)
```